Documentation/design-docs/tiered-compilation.md

```diff
@@ -0,0 +1,95 @@
+
+2. At startup a timer is initiated with a 100ms timeout. If any Tier0 jitting occurs while the timer is running then it is reset. If the timer completes without any Tier0 jitting then, and only then, is call counting allowed to commence. This means a method could be called 1000 times in the first 100ms, but the timer will still need to expire and have the method called 30 more times before it is eligible for Tier1. The reason for the timer is to measure whether or not Tier0 jitting is still occuring, which is a heuristic to measure whether or not the application is still in its startup phase. Before adding the timer we observed that both the call counter and background threads compiling Tier1 code versions were slowing down the foreground threads trying to complete startup, and this could result in losing all the startup performance wins from Tier0 jitting. By delaying until after 'startup' the Tier0 code is left running longer, but that was nearly always a better performing outcome than trying to replace it with Tier1 code too eagerly. 
+
+After these two conditions are satisfied the method is placed in a queue for Tier1 compilation, compiled on a background thread, and then the Tier1 version is made active.
+
+Known Issues (some of which have already generated policy changes post-2.1 RTM): 
+
+1. The call counter may not adequately address cases where methods are hot by virtue of containing loops, even if they aren't invoked many times.
+2. The fixed time value on the timer doesn't scale depending on hardware or process resource constraints.
+3. Once the 100ms elapses with no Tier0, the timer never restarts. Some applications go through multiple iterations of startup-like behavior. For example a server may have an initial burst of activity to listen for requests and then go idle. Then a request arrives more than 100ms later, triggering the first execution of all the request handling code which begins jitting heavily once again.
+4. Its not clear if Tier0 jitting is the best indicator for startup-like behavior, or the best indicator that background jitting would contend for resources with the foreground thread.
+
+Despite the anticipated shortcomings, the 2.1 RTM policy is a surprisingly decent heuristic for many scenarios we measured.
+
+Implementation
+==============
+
+The majority of the implementation can be located in [tieredcompilation.h](../../src/vm/tieredcompilation.h), and [tieredcompilation.cpp](../../src/vm/tieredcompilation.cpp)
+
+The call counter is implemented in [callcounter.h](../../src/vm/callcounter.h), and [callcounter.cpp](../../src/vm/callcounter.cpp)
+
+The policy that determines which methods are eligible for tiering is implemented in MethodDesc::IsEligibleForTieredCompilation, located in [method.hpp](../../src/vm/method.hpp)
```

Comment:
Should `MethodDesc::IsEligibleForTieredCompilation` be enclosed with backticks? To show it as code...